Add repeat Spark function

zhli1142015 opened the pull request:
Doc: https://spark.apache.org/docs/latest/api/sql/#repeat
Code:https://github.com/apache/spark/blob/257a7883f2150e037eb05f8c7a111184103ad9a1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L1885
BTW, `repeat` is used for other fucntions for Presto, so register this with a different name.
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp#L160

zhli1142015 commented:
cc @PHILO-HE  and @rui-mo , thanks.

PHILO-HE commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -459,6 +459,9 @@ void registerFunctions(const std::string& prefix) {
       Array<Array<Generic<T1>>>>({prefix + "flatten"});
 
+  registerFunction<RepeatFunction, Varchar, Varchar, int32_t>(
+      {prefix + "string_repeat"});
```
I think we can also add the name "repeat" as function alias, as it is Spark's function name. Temporarily we call "string_repeat" in Gluten to avoid conflict issue with presto.

PHILO-HE commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,12 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT overlay('Spark SQL', 'tructured', 2, 4); -- "Structured SQL"
         SELECT overlay('Spark SQL', '_', -6, 3); -- "_Sql"
 
+.. spark:function:: string_repeat(str, n) -> varchar
```
string_repeat -> repeat
And then document "string_repeat" separately to just indicate it's an alias for "repeat" function.

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,17 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+
+        SSELECT repeat('123', 2); -- 123123
+
+.. function:: string_repeat(str, n) -> varchar
```
nit: function -> spark:function?

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,17 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    Returns the string which repeats the given string value ``n`` times. 
+    ``n`` must be less or equal than 10000. ::
+
+        SSELECT repeat('123', 2); -- 123123
```
nit: SSELECT -> SELECT

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,17 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+.. spark:function:: repeat(str, n) -> varchar
+
+    Returns the string which repeats the given string value ``n`` times. 
+    ``n`` must be less or equal than 10000. ::
```
Why do we add this limitation? I tried in Spark and found a larger value for n was allowed.

rui-mo replied:
I would be better to document the behavior when n is negative.

zhli1142015 replied:
Updated the doc.
Spark fuzzer test failed with exit code 137 which indicates out of memory error. I thought this is related to this parameter.
 I think 10000 times is big enough, not sure if larger vaule would be use in real cases. Or we should disable this fucntion in fuzzer test, please suggest.

PHILO-HE replied:
@zhli1142015, could you get the input size when OOM happens? Maybe, we should limit the value of `input_size * times`. If this value is greater than some threshold, let the execution fail with exception.

PHILO-HE replied:
Or let's wait for the review from @mbasmanova.

PHILO-HE replied:
@zhli1142015, I see. The limit is an empirical value. I'm not sure how to handle. @mbasmanova, do you have any idea? Should we just let fuzzer test ignore the out-of-memory failure instead of introducing an empirical limit?

rui-mo replied:
I wonder if we can introduce a flag like `FLAGS_repeat_times` and set it as an acceptable value when running the fuzzer test to avoid OOM.

mbasmanova replied:
@zhli1142015 You can reproduce fuzzer failures using --seed 2932607463. You can then debug these using standard tools.

Overall, it would be nice to add a limit on the size of the output string.

https://github.com/facebookincubator/velox/pull/10218#discussion_r1652969475

rui-mo commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -448,6 +449,19 @@ TEST_F(StringTest, overlayVarbinary) {
+  EXPECT_EQ(stringRepeat("abab", 0), "");
+  EXPECT_EQ(stringRepeat("123\u6570", 2), "123\u6570123\u6570");
+  VELOX_ASSERT_USER_THROW(
+      stringRepeat("hh", 10001), "Repeat times is too big.");
```
Can we add tests for negative times and empty input string?

zhli1142015 replied:
Added, thanks

rui-mo reviewed (commented):
Thanks!

PHILO-HE commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,18 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+
+        SELECT repeat('123', 2); -- 123123
+
+.. spark:function:: string_repeat(str, n) -> varchar
```
Please also put this in alphabetical order, though it's just for documenting alias.

PHILO-HE commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+
+    Returns the string which repeats the given string value ``n`` times. 
+    ``n`` must be less or equal than 10000.
+    If ``n`` is less or equal than 0, empry string is returned. ::
```
-> If \``n\`` is less than 0 or equal to 0,
Or
-> If \``n <= 0\``,

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT overlay('Spark SQL', 'tructured', 2, 4); -- "Structured SQL"
         SELECT overlay('Spark SQL', '_', -6, 3); -- "_Sql"
 
+.. spark:function:: repeat(str, n) -> varchar
```
nit: use the full name string to replace str.

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
+.. spark:function:: repeat(str, n) -> varchar
+
+    Returns the string which repeats the given string value ``n`` times. 
```
nit: "string value" could be changed as \``string`` to refer to the input variable.

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1181,6 +1181,34 @@ struct FindInSetFunction {
+    if (inputSize == 0) {
+      result.resize(0);
+      return;
+    }
```
It seems the two branches could be combined as below.

> if (times <= 0 || input.size() == 0) {
result.resize(0);
return;
}

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1181,6 +1181,34 @@ struct FindInSetFunction {
+      result.resize(0);
+      return;
+    }
+    int32_t newSize = velox::checkedMultiply<int32_t>(inputSize, times);
```
In Spark, integer overflow exception is thrown if numBytes * times exceeds the max value of int. Could we add a test to assert the exception?

zhli1142015 replied:
Added, thanks.

rui-mo reviewed (commented):
Thanks.

jinchengchenghh commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -459,6 +459,9 @@ void registerFunctions(const std::string& prefix) {
       Array<Array<Generic<T1>>>>({prefix + "flatten"});
 
+  registerFunction<RepeatFunction, Varchar, Varchar, int32_t>(
+      {prefix + "repeat", "string_repeat"});
```
"string_repeat" -> prefix + "string_repeat"

zhli1142015 replied:
Good catch, thanks

jinchengchenghh commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -448,6 +449,24 @@ TEST_F(StringTest, overlayVarbinary) {
+TEST_F(StringTest, repeat) {
+  const auto stringRepeat = [&](const std::optional<std::string>& str,
+                                const std::optional<int32_t>& times) {
+    return evaluateOnce<std::string>("repeat(c0, c1)", str, times);
```
should also test string_repeat function name

zhli1142015 replied:
Added, thanks

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -220,6 +228,10 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
+.. spark:function:: string_repeat(str, n) -> varchar
+
+    This is an alias for ``repeat(str, n)``.
```
nit: It would be nice to make 'repeat' a link using :spark:func:.

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+
+    Returns the string which repeats ``string`` ``n`` times. 
+    ``n`` must be less than or equal to 10000.
+    If ``n`` is less than or equal to 0, empry string is returned. ::
```
typo: empry -> empty

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -141,6 +141,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+.. spark:function:: repeat(string, n) -> varchar
+
+    Returns the string which repeats ``string`` ``n`` times. 
+    ``n`` must be less than or equal to 10000.
```
It would be nice to add a limit on the total size of the result string. Maybe, something like in lpad / rpad. Does Spark has a limit?

```
template <bool lpad, bool isAscii, typename TOutString, typename TInString>
FOLLY_ALWAYS_INLINE void pad(
    TOutString& output,
    const TInString& string,
    const int64_t size,
    const TInString& padString) {
  static constexpr size_t padMaxSize = 1024 * 1024; // 1MB
```

zhli1142015 replied:
> Does Spark has a limit?

I think in spark the limit is int32.max, I think it's a too big value to fuzzer test.
https://github.com/apache/spark/blob/a50b30d7a02bf45f1ddb8db0be6779b1441e1d4d/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L870

mbasmanova replied:
> I think it's a too big value to fuzzer test.

More importantly this is too large for production. How about we add 1 MB limit to the size of the result string?

zhli1142015 replied:
I'm ok for 1 MB limit, @rui-mo , is this ok to you?

rui-mo replied:
@zhli1142015 Yes, it makes sense to me.

zhli1142015 replied:
Updated, thanks.

zhli1142015 commented:
@rui-mo , do you have more comments? Thanks.

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1181,6 +1181,35 @@ struct FindInSetFunction {
 };
 
+template <typename T>
+struct RepeatFunction {
```
nit: add some comments for this function. Some example in the replace function:

> /// replace(input, replaced) -> varchar
///
///     Removes all instances of ``replaced`` from ``input``.
///     If ``replaced`` is an empty string, returns the original ``input``
///     string.

zhli1142015 replied:
added, thanks

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1181,6 +1181,41 @@ struct FindInSetFunction {
+/// repeat(input, times) -> varchar
+/// string_repeat(input, times) -> varchar
+///
+///    Returns the string which repeats ``input`` ``times`` times.
```
nit: 
> repeat(string, n)
string_repeat(string, n) -> varchar

> Returns the string which repeats the give ``string`` ``n`` times.

PHILO-HE commented:
@zhli1142015, one minor comment. I note you say "repeat is used for other fucntions for Presto, so register this with a different name.". We have removed presto's scalar function registration from Gluten. There should be no conflict issue if we use the registered name "repeat" in Gluten. So we can remove the unnecessary alias name `string_repeat`.

zhli1142015 commented:
> @zhli1142015, one minor comment. I note you say "repeat is used for other fucntions for Presto, so register this with a different name.". We have removed presto's scalar function registration from Gluten. There should be no conflict issue if we use the registered name "repeat" in Gluten. So we can remove the unnecessary alias name `string_repeat`.

Thanks for sharing this, updated.

PHILO-HE reviewed (approved):
Looks nice!
